openssl-dynamic/src/main/c/sslutils.c

```diff
@@ -482,11 +568,89 @@ int SSL_CTX_use_certificate_chain_bio(SSL_CTX *ctx, BIO *bio,
               && ERR_GET_REASON(err) == PEM_R_NO_START_LINE)) {
             return -1;
         }
-        while (ERR_get_error() > 0) ;
+        ERR_clear_error();
     }
     return n;
 }
 
+int SSL_use_certificate_chain_bio(SSL *ssl, BIO *bio,
+                                  int skipfirst)
+{
+#if !defined(OPENSSL_IS_BORINGSSL) && OPENSSL_VERSION_NUMBER < 0x1000200fL
+    // Only supported on boringssl or openssl 1.0.2+
+    return -1;
+#else
+    X509 *x509;
+    unsigned long err;
+    int n;
+    STACK_OF(X509) *chain;
+
+    /* optionally skip a leading server certificate */
+    if (skipfirst) {
+        if ((x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL)) == NULL) {
+            return -1;
+        }
+        X509_free(x509);
+    }
+
+    /* create new extra chain by loading the certs */
+    n = 0;
+
+    while ((x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL)) != NULL) {
+        if (SSL_add0_chain_cert(ssl, x509) != 1) {
+            X509_free(x509);
+            return -1;
+        }
+        n++;
+    }
+    /* Make sure that only the error is just an EOF */
+    if ((err = ERR_peek_error()) > 0) {
+        if (!(   ERR_GET_LIB(err) == ERR_LIB_PEM
+              && ERR_GET_REASON(err) == PEM_R_NO_START_LINE)) {
+            return -1;
+        }
+        ERR_clear_error();
+    }
+    return n;
+#endif
+}
+
+X509 *load_pem_cert_bio(tcn_pass_cb_t *cb_data, const BIO *bio)
+{
+    X509 *cert = NULL;
+    if (cb_data == NULL)
+        cb_data = &tcn_password_callback;
+    cert = PEM_read_bio_X509_AUX((BIO*) bio, NULL,
+                (pem_password_cb *)SSL_password_callback,
+                (void *)cb_data);
+    if (cert == NULL &&
+       (ERR_GET_REASON(ERR_peek_last_error()) == PEM_R_NO_START_LINE)) {
+        ERR_clear_error();
+        BIO_ctrl((BIO*) bio, BIO_CTRL_RESET, 0, NULL);
+        cert = d2i_X509_bio((BIO*) bio, NULL);
+    }
+    return cert;
+}
+
+EVP_PKEY *load_pem_key_bio(tcn_pass_cb_t *cb_data, const BIO *bio)
+{
+    EVP_PKEY *key = NULL;
+    int i;
+
+    if (cb_data == NULL)
+        cb_data = &tcn_password_callback;
```

Comment:
same issue for `load_pem_cert_bio` ... i wonder if we should be setting `c->cb_data = &tcn_password_callback;` and nulling put a `null` in the first element event if `password` is  `null` in `sslcontext::setCertificateBio`?